capsules/src/nonvolatile_to_pages.rs

```diff
@@ -0,0 +1,263 @@
+                        self.buffer.replace(buffer);
+                        // Increment all buffer pointers and state.
+                        self.remaining_length.set(self.remaining_length.get() - len);
+                        self.address.set(self.address.get() + len);
+                        self.buffer_index.set(buffer_index + len);
+                        self.driver.read_page(self.address.get() / page_size, pagebuffer);
+                    }
+                });
+            }
+            State::Write => {
+                // We did a read because we're not page aligned on either or
+                // both ends.
+                self.buffer.take().map(move |buffer| {
+                    let page_size = pagebuffer.as_mut().len();
+                    // This will get us our offset into the page.
+                    let page_index = self.address.get() % page_size;
+                    // Length is either the rest of the page or how much we have left.
+                    let len = cmp::min(page_size - page_index, self.remaining_length.get());
+                    // And where we left off in the user buffer.
+                    let buffer_index = self.buffer_index.get();
+                    // Which page we read and which we are going to write back to.
+                    let page_number = self.address.get() / page_size;
+
+                    // Copy what we read from the page buffer to the user buffer.
+                    for i in 0..len {
+                        pagebuffer.as_mut()[page_index + i] = buffer[buffer_index + i];
+                    }
+
+                    // Do the write.
+                    self.buffer.replace(buffer);
+                    self.remaining_length.set(self.remaining_length.get() - len);
+                    self.address.set(self.address.get() + len);
+                    self.buffer_index.set(buffer_index + len);
+                    self.driver.write_page(page_number, pagebuffer);
+                });
+            }
+            _ => {}
+        }
+
+    }
+
+    fn write_complete(&self, pagebuffer: &'static mut F::Page, _error: hil::flash::Error) {
+        // After a write we could be done, need to do another write, or need to
+        // do a read.
+        self.buffer.take().map(move |buffer| {
+            let page_size = pagebuffer.as_mut().len();
+
+            if self.remaining_length.get() == 0 {
+                // Done!
+                self.pagebuffer.replace(pagebuffer);
+                self.state.set(State::Idle);
+                self.client.get().map(move |client| {
+                    client.write_done(buffer, self.length.get())
+                });
+            } else if self.remaining_length.get() >= page_size {
+                // Write an entire page!
+                let buffer_index = self.buffer_index.get();
+                let page_number = self.address.get() / page_size;
+
+                // Copy data into page buffer.
+                for i in 0..page_size {
```

Comment:
could iterate here instead of doing an index:

```rust
for s,d in pagebuffer.as_mut().iter_mut().fold(buffer.iter()) {
  *s = d;
}
```

Which, by construction, can't have an index out of bounds error